core/controllers/reader_test.py

```diff
@@ -205,6 +248,103 @@ def test_huge_answers(self):
 
 
+class ReaderClassifyTests(test_utils.GenericTestBase):
+    """Test reader.classify using the sample explorations.
+
+    Since the end to end tests cover correct classification,
+    ReaderClassifyTests is only checking which of the hard/soft/classifier
+    rules triggered on an input.
+    """
+
+    def setUp(self):
+        super(ReaderClassifyTests, self).setUp()
+        self.init_classify_inputs('16')
+
+    def init_classify_inputs(self, exploration_id):
+        exp_services.delete_demo(exploration_id)
+        exp_services.load_demo(exploration_id)
+
+        self.EXP_ID = exploration_id
+        self.EXP_STATE = (
+            exp_services.get_exploration_by_id(exploration_id).states['Home'])
+
+    def determine_triggering_rule(self, answer):
+        response = reader.classify(self.EXP_ID, self.EXP_STATE, answer,
+            {'answer': answer})
+        return response['matched_rule_type']
+
+    def test_hard_triggers(self):
+        """All these responses trigger the hard classifier.
+
+        Note: Any response beginning with 'hardrule' is a hard rule
+        triggering.
+        """
+        self.assertEquals(
+            self.determine_triggering_rule('permutations'),
+            'hard')
+        self.assertEquals(
+            self.determine_triggering_rule('hardrule0'),
+            'hard')
+        self.assertEquals(
+            self.determine_triggering_rule('hardrule3'),
+            'hard')
+        self.assertEquals(
+            self.determine_triggering_rule('hardrule1254'),
+            'hard')
+        self.assertEquals(
+            self.determine_triggering_rule('exit'),
+            'hard')
+
+    def test_soft_triggers(self):
+        """All these responses trigger the soft classifier."""
+        self.assertEquals(
+            self.determine_triggering_rule('Combination 3 x 2 x 1'),
+            'soft')
+        self.assertEquals(
+            self.determine_triggering_rule('bcse combinations'),
+            'soft')
+        self.assertEquals(
+            self.determine_triggering_rule('Because the answer is 3!'),
+            'soft')
+        self.assertEquals(
+            self.determine_triggering_rule('3 balls time two time one'),
+            'soft')
+        self.assertEquals(
+            self.determine_triggering_rule('try all possible combinations'),
+            'soft')
+        self.assertEquals(
+            self.determine_triggering_rule('rby, ryb, bry, byr, ybr, yrb'),
+            'soft')
+        self.assertEquals(
+            self.determine_triggering_rule('I dunno!'),
+            'soft')
+        self.assertEquals(
+            self.determine_triggering_rule('I guessed.'),
+            'soft')
+
+    def test_classify_triggers(self):
+        """All these responses trigger the string classifier."""
+        with self.swap(feconf, 'ENABLE_STRING_CLASSIFIER', True):
+            self.assertEquals(
+                self.determine_triggering_rule('it\'s a permutation of 3 '
```

Comment:
Nit: The string should begin on the next line (as mentioned earlier). Ditto for other occurrences in this file.